### PR TITLE
Structured data/minirocket

### PR DESCRIPTION
**Pull request by ThomasUyttenhove**
README and notebook for Structured Data quick tip: time series classification with MINIROCKET

### CODE REVIEW

**Comment by julienschuermans** [comment]
preview the notebook here:
https://colab.research.google.com/github/ml6team/quick-tips/blob/structured_data%2Fminirocket/structured_data/2021_05_25_minirocket_timeseries_classification/minirocket_timeseries_classification.ipynb

**Review by julienschuermans — COMMENTED** [review]
Great work, thanks Thomas! :)
I added some minor remarks. Looks very good overall.

File: structured_data/2021_05_25_minirocket_timeseries_classification/README.md [hunk 1]
```diff
@@ -0,0 +1,9 @@
+# Time series: classification with MINIROCKET
+For time series classification, we usually rely on some sort of feature extraction combined with a traditional classifier. While the _tsfresh_ package is often preferred for the former task, it becomes exponentially slow for growing dataset sizes thus making it cumbersome or even impossible to use in real-life use cases. 
+
+In a publication preprint by _A. Dempster et al._, [_MINIROCKET: A Very Fast (Almost) Deterministic Transform
+for Time Series Classification_](https://arxiv.org/pdf/2012.08791.pdf), an alternative kernel-based feature transformation method is proposed. It can extract features in both CPU and GPU environments at a fraction of _tsfresh_'s execution time (especially for large datasets) while enabling on par performance at the cost of feature interpretability. Additionally, it generally offers better performance than Fourier transformed (_fft_/_stft_) features.
```
Comment (FlorentijnD) [inline structured_data/2021_05_25_minirocket_timeseries_classification/README.md:5]: We compare the speed of minirocket to tsfresh, and talk about "general performance" when comparing it to Fourier. Could we also add speed in comparison to Fourier and general performance wrt tsfresh?

Comment (FlorentijnD) [inline structured_data/2021_05_25_minirocket_timeseries_classification/README.md:5]: We compare the speed of minirocket to tsfresh, and talk about "general performance" when comparing it to Fourier. Could we also add speed in comparison to Fourier and general performance wrt tsfresh?

**Review by FlorentijnD — COMMENTED** [review]
- things that are not entirely clear to me : is MiniRocket_fastai() equal to the RidgeClassifierCV() in combination with MiniRocketMultivariate_sktime()/MiniRocket_sktime()?
- Why do we need the helper functions exactly? Why are they not implemented in the libraries themselves? (Can we make a PR out of this to the repo's?)
- I have the same remark as Vincent had in chat; the training/validation loss curve seems like the system is overfitting slightly :)

File: structured_data/2021_05_25_minirocket_timeseries_classification/README.md (hunk 1 above)
Comment (ThomasUyttenhove) [reply structured_data/2021_05_25_minirocket_timeseries_classification/README.md:5]: Good remark! The description actually does already mention that _tsfresh_ and _MINIROCKET_ have on par performance (it really depends on the dataset which one comes out on top. I reworded this section to give a clearer overview.

**Comment by ThomasUyttenhove** [comment]
@FlorentijnD 
> Thanks Thomas! :D
> 
> things that are not entirely clear to me :
> 
> * is MiniRocket_fastai() equal to the RidgeClassifierCV() in combination with MiniRocketMultivariate_sktime()/MiniRocket_sktime()?
> * Why do we need the helper functions exactly? Why are they not implemented in the libraries themselves? (Can we make a PR out of this to the repo's?)
> * I have the same remark as Vincent had in chat; the training/validation loss curve seems like the system is overfitting slightly :)

To clear things up:
* MiniRocket_fastai() is the renamed (during import) version of `tsai` MiniRocketClassifier to avoid namespace overlap between the `tsai` and `sktime` packages. This function provides their implementation of the _MINIROCKET_ feature transform and a ridge classifier, while `sktime` requires you to combine both yourself but does give you the freedom to incorporate it in an `sklearn` pipeline. (This is indeed just a wrapper around the MiniRocketMultivariate_sktime() and RidgeClassifierCV() functions, but in my experience a tad slower than when you would implement these yourself!)
* The helper functions are source code from the `tsai` and `sktime` packages that was edited to on the one hand calculate "relevant" `tsfresh` features (as a form of feature selection) and on the other hand to work around the fact that the UEA/UCR repository does not provide each dataset in _.ts_ files but often does provide _.arff_ files. The former could indeed be an interesting extension/option in the existing `tsai` features, but the last one is more of a dirty hack (that does not work all the time).
* I tweaked the parameters of the GPU implementation to better converge!

**Review by julienschuermans — APPROVED** [review]
Great work @ThomasUyttenhove !
Looks good to me!

**Review by Fematich — APPROVED** [review]
Thx! very clear LGTM.

2 suggestions for next time:
- The notebook is clear, however it's very long due to a lot of helper code. Not sure if the benefit of a single notebooks outweights the benefit of having some external helper functions to keep the notebook clean? 
- You're making a nice trade-off comparision between speed-accuracy for 2 datasets & multiple approaches (2 times ts-fresh, 2 times minirocket & 1 time FFT). I was looking for a summary table or a scatter plot with the 2 axes (inference speed & accuracy) for these approaches as the one picture that summarizes the whole notebook :-).

**Comment by ThomasUyttenhove** [comment]
@Fematich 

> Thx! very clear LGTM.
> 
> 2 suggestions for next time:
> 
> * The notebook is clear, however it's very long due to a lot of helper code. Not sure if the benefit of a single notebooks outweights the benefit of having some external helper functions to keep the notebook clean?
> * You're making a nice trade-off comparision between speed-accuracy for 2 datasets & multiple approaches (2 times ts-fresh, 2 times minirocket & 1 time FFT). I was looking for a summary table or a scatter plot with the 2 axes (inference speed & accuracy) for these approaches as the one picture that summarizes the whole notebook :-).

Thanks for the feedback! :)
- I agree with the notebook being quite lengthy; for future notebooks we will look into how we can easily integrate .py files from the repository in the Colab environment to avoid this!
- Great point! I have added two tables tie the whole notebook together.
